extensions/notebook/src/book/remoteBookController.ts

```diff
@@ -0,0 +1,261 @@
+				let releases = JSON.parse(body);
+				let bookReleases: IReleases[] = [];
+				if (releases) {
+					let keys = Object.keys(releases);
+					keys = keys.filter(key => {
+						let release = {} as IReleases;
+						try {
+							release.remote_path = url;
+							release.tag_name = releases[key].tag_name;
+							release.tarURL = new URL(releases[key].tarball_url);
+							release.zipURL = new URL(releases[key].zipball_url);
+						}
+						catch (error) {
+							throw (error);
+						}
+						bookReleases.push(release);
+					});
+				}
+				resolve(bookReleases);
+			});
+		});
+	}
+}
+
+export interface IReleases {
+	remote_path: URL;
+	tag_name: string;
+	zipURL: URL;
+	tarURL: URL;
+}
+
+export abstract class RemoteBook {
+	protected _book_name: string;
+	protected _version: string;
+	protected _lang_code: string;
+	protected _local_path: URL;
+	protected _tarURL: URL;
+	protected _zipURL: URL;
+	protected outputChannel: vscode.OutputChannel;
+	protected apiWrapper: ApiWrapper;
+
+	constructor(public remote_path: URL) {
+		this.remote_path = remote_path;
+		this.apiWrapper = new ApiWrapper();
+		this.outputChannel = this.apiWrapper.createOutputChannel(msgTaskName);
+	}
+
+	public async abstract createLocalCopy(): Promise<string[]>;
+
+	public openRemoteBook(selectedBook: string) {
+		vscode.commands.executeCommand('bookTreeView.openBook', selectedBook, false, undefined);
+	}
+
+	public setLocalPath() {
+		// Save directory on User directory
+		if (vscode.workspace.workspaceFolders !== undefined) {
+			// Get workspace root path
+			let folders = vscode.workspace.workspaceFolders;
+			try {
+				this._local_path = new URL(folders[0].uri.fsPath);
+			}
+			catch (error) {
+				throw (error);
+			}
+		} else {
+			//If no workspace folder is opened then path is Users directory
+			try {
+				this._local_path = new URL(utils.getUserHome());
+			}
+			catch (error) {
+				throw (error);
+			}
+		}
+	}
+
+	public loadBooks(dir: string, books: string[]): string[] {
+		fs.readdirSync(dir).forEach(file => {
+			let fullPath = path.join(dir, file);
+			if (fs.lstatSync(fullPath).isDirectory()) {
+				if (isBook(fullPath)) {
+					books.push(fullPath);
+				}
+				return this.loadBooks(fullPath, books);
+			}
+			return [];
+		});
+		return books;
+	}
+}
+
+class SharedRemoteBook extends RemoteBook {
+	constructor(public remote_path: URL) {
+		super(remote_path);
+	}
+
+	public async createLocalCopy(): Promise<string[]> {
+		//compress it
+		// copy it
+		// to local path
+		return [];
+	}
+}
+
+export class GitHubRemoteBook extends RemoteBook {
+	constructor(public remote_path: URL, public zipURL: URL, public tarURL: URL) {
+		super(remote_path);
+		this._zipURL = zipURL;
+		this._tarURL = tarURL;
+	}
+
+	public async createLocalCopy(): Promise<string[]> {
+		this.outputChannel.show(true);
+		this.outputChannel.appendLine(msgRemoteBookDownloadProgress);
+		return new Promise((resolve, reject) => {
+			this.setLocalPath();
+			let re = /\//g;
```

Comment:
Make this a const in the root of the file - you don't need to recreate it each time here.